### PR TITLE
Add `download-manifest` command; file globbing for `download`

### LINKED ISSUE #123: Add cli command to download files specified in manifest file constructed by the data browser
The Data Browser facet search will construct a manifest file specifying UUIDs and timestamps for files the user wants to download.
We need to provide a "hca download <manifest_file>" cli command to do the actual download (including retries).

### PR DESCRIPTION
**Pull request by hannes-ucsc**
Fixes #123

Adds the ability to download files from a DSS instance given a tab-separated file with references to said files. Intuitively, one would think that the TSV only needs to contain a list of file FQIDs. So why did I not go this route? Here is the train of thought that led to the current implementation:

1) From a usability perspective it is desirable that a downloaded file is assigned the same name that the file had at submission time.

2) That file name is tracked in the bundle manifest, not the file object. Given a file's FQID one would also need the FQID of a bundle that references the file. There could be more than one bundle referencing a file. The file name could be different between those bundles. To be unambiquous, the manifest has to contain two FQIDs per row, that of  of the file and that of a bundle containing the file and assigning it a name.

3) There's another catch: Two distinct files in in two distinct bundles could be assigned the same name. In order to download them both into a single directory, one would need to apply some disambiguation to the file's name. Appending the file's FQID comes to mind, or placing the files into separate sub-directories. The former seems inconvenient. For the latter, the obvious choice for naming the subdirectory is the bundle UUID. The resulting directory layout is identical to the one produced by the existing single-bundle `download` command. Using it for both `download` and `download-manifest` would enable a synergy between them: the user could augment the result of a manifest download by later running `download` on individual bundles.

4) To achieve this, we could simply extend the existing `download` command to accept a set of file FQIDs that limits what files are being downloaded. While this would be sufficient to support the manifest download, we can kill two birds with one stone by using file names instead of identifiers to restrict the download. That way, the user can say `hca download --bundle-uuid $UUID --version $version --files SRR2967608_1.fastq.gz`.

5) With just a tiny bit more code we can extend that mechanism from file names to shell patterns (globs). Now the user can say `hca download --bundle-uuid $UUID --version $version --files *.fastq.gz`. That is a big boost in convenience and control. Globs also make it easy to assign a backwards-compatible default value to the new `--files` parameter: `*`. That's intuitive and self-explanatory. If the `--files` parameter only accepted literal file names, the default would have to be a less intuitive value, like `None` or the empty string.

6) In order to safely use file names instead of FQIDs for selecting which file to download from a bundle, we need to make sure that those file names are actually unique within the bundle. However, this uniqueness constraint is not just requirement for this feature. It is in fact a requirement for bundle downloads in general. If we don't prevent conflicting file names in a bundle, the existing `download` command will happily overwrite files of the same name, or worse, mix data from two remote files into one local file, in the case of a resumed download. Once we address that, the triple `(bundle_uuid, bundle_version, file_name)` is a safe and stable identifier. There is a ticket open in the data-store project that aims to enforce uniqueness of file names in newly submitted bundles. This PR adds code to handle any existing bundles that may have file name collisions. There is a twist with respect to case-insensitive file systems but the respective commit handles that.

Related PRs:

https://github.com/HumanCellAtlas/data-store/issues/1268
https://github.com/HumanCellAtlas/data-store/issues/1331

### CODE REVIEW

File: hca/dss/__init__.py [hunk 1]
```diff
@@ -21,27 +26,62 @@ class DSSClient(SwaggerClient):
...
             dest_name = bundle_uuid
 
         bundle = self.get_bundle(uuid=bundle_uuid, replica=replica, version=version if version else None)["bundle"]
 
-        if not os.path.isdir(dest_name):
-            os.makedirs(dest_name)
-
+        files = {}
         for file_ in bundle["files"]:
+            # The file name collision check is case-insensitive even if the local file system we're runnign on is
```
Comment (kislyuk) [inline hca/dss/__init__.py]: *running

File: requirements-dev.txt [hunk 2]
```diff
@@ -5,4 +5,5 @@ pyyaml
 responses
 backports.tempfile
 mock
+unittest2
```
Comment (kislyuk) [inline requirements-dev.txt]: Why is this necessary? Is there something missing from the Python 2.7 stdlib unittest?

File: hca/util/compat.py [hunk 3]
```diff
@@ -1,5 +1,29 @@
...
+import os
+import re
+import sys
 
 USING_PYTHON2 = True if sys.version_info < (3, 0) else False
+
+if USING_PYTHON2:
+    # https://github.com/python/cpython/blob/master/Lib/glob.py#L142
+    magic_check = re.compile('([*?[])')
+    magic_check_bytes = re.compile(b'([*?[])')
```
Comment (kislyuk) [inline hca/util/compat.py]: Consider making these "private" by prefixing with underscore.

File: requirements-dev.txt (hunk 2 above)
Comment (hannes-ucsc) [reply requirements-dev.txt]: Python3 added TestCase.subTest() which is very useful and this backport has it too.

File: hca/util/compat.py (hunk 3 above)
Comment (hannes-ucsc) [reply hca/util/compat.py]: Fixed
